KZOO-516: added logic to handle crashing for invalid metaflow schema

sanjeevie6353 opened the pull request:
Previously, the system crashed when an incorrect schema body is sent in the API request when trying to activate call recording.  The system sent an a HTTP 500 error. 

This new change will handle that error and will send a HTTP 400 error, stating a client error.

sanjeevie6353 commented:
Task details 

ticket - https://oomacorp.atlassian.net/browse/KZOO-516

test cases - 
[https://oomacorp.atlassian.net/projects/KZOO?selectedItem=com.atlassian.plugins.atlassian-connect-plugin:com.kanoah.test-manager__main-project-page#!/v2/testCase/215710134](url)
[https://oomacorp.atlassian.net/projects/KZOO?selectedItem=com.atlassian.plugins.atlassian-connect-plugin:com.kanoah.test-manager__main-project-page#!/v2/testCase/215710134](url)

Scenario - 
Previously, the system crashed when an incorrect schema body is sent in the API request when trying to activate call recording. The system sent an a HTTP 500 error.

```
|a174481a93da3144a1e0ed702fd87546|kz_json_schema_extensions:72(<0.6880097.0>) validating metaflow action 'undefined' with data {[{<<"module">>,<<"record_call">>},{<<"data">>,{[{<<"action">>,<<"start">>}]}}]}
2024-08-21T19:40:51.078011+00:00 aio 2600hz[13996]: |a174481a93da3144a1e0ed702fd87546|kazoo_bindings:696(<0.6880097.0>) excepted: error: badarg
2024-08-21T19:40:51.092562+00:00 aio 2600hz[13996]: |a174481a93da3144a1e0ed702fd87546|kz_log:180(<0.6880097.0>) stacktrace: 
2024-08-21T19:40:51.117567+00:00 aio 2600hz[13996]: |a174481a93da3144a1e0ed702fd87546|kz_log:187(<0.6880097.0>) st: kz_json_schema_extensions:extra_validation/3 at (73)
2024-08-21T19:40:51.144844+00:00 aio 2600hz[13996]: |a174481a93da3144a1e0ed702fd87546|kz_log:187(<0.6880097.0>) st: lists:foldl_1/3 at (1599)
2024-08-21T19:40:51.197040+00:00 aio 2600hz[13996]: |a174481a93da3144a1e0ed702fd87546|kz_log:187(<0.6880097.0>) st: jesse_validator_draft4:check_value/4 at (276)
2024-08-21T19:40:51.222935+00:00 aio 2600hz[13996]: |a174481a93da3144a1e0ed702fd87546|kz_log:187(<0.6880097.0>) st: lists:foldl_1/3 at (1599)
2024-08-21T19:40:51.250662+00:00 aio 2600hz[13996]: |a174481a93da3144a1e0ed702fd87546|kz_log:187(<0.6880097.0>) st: jesse_validator_draft4:check_properties/3 at (410)
2024-08-21T19:40:51.265615+00:00 aio 2600hz[13996]: |a174481a93da3144a1e0ed702fd87546|kz_log:187(<0.6880097.0>) st: jesse_validator_draft4:check_value/3 at (94)
2024-08-21T19:40:51.281259+00:00 aio 2600hz[13996]: |a174481a93da3144a1e0ed702fd87546|kz_log:187(<0.6880097.0>) st: jesse_schema_validator:validate/3 at (46)
2024-08-21T19:40:51.295650+00:00 aio 2600hz[13996]: |a174481a93da3144a1e0ed702fd87546|kz_log:187(<0.6880097.0>) st: jesse:validate_with_schema/3 at (273)
2024-08-21T19:40:51.308572+00:00 aio 2600hz[13996]: |a174481a93da3144a1e0ed702fd87546|api_util:1350(<0.6880097.0>) validating data failed
```

This new change will handle that error and will send a HTTP 400 error, stating a client error.

manushiperera commented:
The code changes seem to successfully mitigate the crash.
And both test cases have successfully passed in my development environment.

jamesaimonetti commented on kazoo_schemas/src/kz_json_schema_extensions.erl:
```diff
@@ -72,7 +72,14 @@ extra_validation(<<"metaflow.data">>, Value, State) ->
+        'undefined' ->
+            jesse_error:handle_data_invalid({'external_error', <<"invalid metaflow schema for module">>}
+                                           ,Value
+                                           ,State);
```
drop `);` to new line (should be same indentation level as opening `(`

sanjeevie6353 replied:
This has been changed to `);` available in a new line.

jamesaimonetti commented on kazoo_schemas/src/kz_json_schema_extensions.erl:
```diff
@@ -72,7 +72,14 @@ extra_validation(<<"metaflow.data">>, Value, State) ->
     Module = jesse_json_path:path(lists:reverse([<<"module">> | Path]), JObj, 'undefined'),
     lager:debug("validating metaflow action '~s' with data ~p", [Module, Value]),
-    validate_module_data(<<"metaflows.", Module/binary>>, Value, State);
+    case Module of
```
Personally I would put the `jesse_json_path:path/3` call here in the case statement, since we don't need `Module` outside of the case clause below. No need to capture it until its needed. You could move the log line into the case clause too (and add a line in the `undefined` clause that the metaflow is missing the action module).

sanjeevie6353 replied:
This change has been added to the PR.

lazedo commented on kazoo_schemas/src/kz_json_schema_extensions.erl:
```diff
@@ -70,9 +70,17 @@ extra_validation(<<"accounts.announcement">>, Value, State) ->
+    case jesse_json_path:path(lists:reverse([<<"module">> | Path]), JObj, 'undefined') of
+        'undefined' ->
+            lager:debug("metaflow action missing in path ~p", [Path]),
+            jesse_error:handle_data_invalid({'external_error', <<"invalid metaflow schema for module">>}
```
```suggestion
            jesse_error:handle_data_invalid({'external_error', <<"missing metaflow module">>}
```

sanjeevie6353 replied:
This change has been added to the PR

Commit:
KZOO-516: added logic to handle crashing for invalid metaflow schema

Previously, the system crashed when an incorrect schema body is sent in the API request when trying to activate call recording. The system sent an a HTTP 500 error.

This new change will handle that error and will send a HTTP 400 error, stating a client error.
